linux-libre: init

alyssais opened the pull request:
###### Motivation for this change

GNU Linux-libre is a linux kernel distribution that has all non-free binary blobs stripped. The project distributes both deblobbed kernel sources, and scripts for each kernel version that perform the deblobbing. Here, I have used the scripts rather than the pre-deblobbed sources, because this allows an arbitrary kernel to be passed in and deblobbed with no extra effort, which seems more in the spirit of Nix and allows deblobbing to be composed with other kernel customisations.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

With the small fix to openafs included in this PR, `linuxPackages-libre` and `linuxPackages_latest-libre` can build all kernel modules that the corresponding non-deblobbed kernels can build. (This does not mean that `nox-review` will pass, because some kernel modules are currently broken on non-deblobbed Linux 4.18 anyway. See #45410.) The fix to openafs does not break any existing kernels where it was previously able to build. I have booted and used a deblobbed kernel successfully, with the exception of some of my laptop's hardware not working, which is to be expected with a deblobbed kernel.

alyssais committed:
linuxPackages*.openafs*: use modDirVersion

`version` doesn't include .0 patch versions, or suffixes (eg -gnu).
For .0 patch versions, this isn't a problem, since the glob takes care
of everything after a `.`. It does, however, prevent openafs installing
for suffixed kernels.

alyssais committed:
linux-libre: init

Mic92 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14210,6 +14210,12 @@ with pkgs;
+  linuxPackages-libre = recurseIntoAttrs (linuxPackagesFor linux-libre);
+  linux-libre = callPackage ../os-specific/linux/kernel/linux-libre.nix {};
+  linuxPackages_latest-libre = recurseIntoAttrs (linuxPackagesFor linux_latest-libre);
+  linux_latest-libre = linux-libre.override { linux = linux_latest; };
```
Do we really need both latest and stable?

alyssais replied:
I did this for consistency with other kernels – hardened and xen_dom0 both have latest and stable. I can definitely see both having an audience.

Mic92 commented on pkgs/os-specific/linux/kernel/linux-libre.nix:
```diff
@@ -0,0 +1,34 @@
+  minor = lib.versions.minor linux.modDirVersion;
+  patch = lib.versions.patch linux.modDirVersion;
+
+in linux.override {
```
For things like kernels, I would like to see a maintainer.

alyssais committed:
linux-libre: add myself as a maintainer

xeji committed:
linux-libre: fix argument list

Building a system with boot.kernelPackages = pkgs.linuxPackages-libre
failed because the expression is called with extra arguments,
See linux-4.14.nix.

xeji commented:
Pushed a trivial fix so we can actually build a system with this kernel. LGTM now.

xeji commented:
![selection_002](https://user-images.githubusercontent.com/36407913/44815821-db981000-abe0-11e8-9ce8-545d7c45477d.png)

Mic92 commented:
I suppose non-gpl third-party modules should be blacklisted for this kernel? https://hydra.nixos.org/eval/1476119?filter=linuxPackages-libre&compare=1476064&full=#tabs-new
